### PR TITLE
simplifier cleanup

### PR DESCRIPTION
**Pull request by kroening**
- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3233db8e8350b5e3f4f1c9f317faa3247462393f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116552129

File: src/util/simplify_expr.cpp [hunk 1]
```diff
@@ -848,38 +848,39 @@ simplify_exprt::simplify_typecast(const typecast_exprt &expr)
...
+    tmp_op1.op() = if_expr.true_case();
+    exprt tmp_op1_result = simplify_dereference(tmp_op1);
 
-    expr=if_exprt(if_expr.cond(), tmp_op1, tmp_op2);
+    auto tmp_op2 = expr;
+    tmp_op2.op() = if_expr.false_case();
+    exprt tmp_op2_result = simplify_dereference(tmp_op2);
 
-    simplify_if(to_if_expr(expr));
+    auto tmp = if_exprt(if_expr.cond(), tmp_op1_result, tmp_op2_result);
```
Comment (tautschnig) [inline src/util/simplify_expr.cpp]: `if_exprt tmp{if_expr.cond(), tmp_op1_result, tmp_op2_result};`

**Review by tautschnig — APPROVED** [review]
Just one nit pick.

File: src/util/simplify_expr.cpp [hunk 2]
```diff
@@ -151,41 +146,35 @@ bool simplify_exprt::simplify_popcount(popcount_exprt &expr)
...
-    to_function_application_expr(expr);
-
-  if(function_app.function().id() != ID_symbol)
-    return true;
+  if(expr.function().id() != ID_symbol)
+    return unchanged(expr);
 
-  const irep_idt func_id =
-    to_symbol_expr(function_app.function()).get_identifier();
+  const irep_idt func_id = to_symbol_expr(expr.function()).get_identifier();
```
Comment (peterschrammel) [inline src/util/simplify_expr.cpp]: `const &`

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2d23e6063edf63ec041114cb48ca26ae1253b6a0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116561140

**Pull request merged**